contrib/epee/include/net/net_helper.h

```diff
@@ -110,25 +111,34 @@ namespace net_utils
 			catch(...) { /* ignore */ }
 		}
 
+		inline void set_ssl(epee::net_utils::ssl_support_t ssl_support = epee::net_utils::e_ssl_support_autodetect, const std::string &private_key = std::string(), const std::string &certificate = std::string(), const std::list<std::string> &allowed_certificates = std::list<std::string>(), bool allow_any_cert = false)
```

Comment:
I have simultaneously been working on updating client SSL stuff here since the proxy code has to touch this very same portion of code. Then I realized that this PR almost certainly changed this function (yup).

I started to use a `boost::variant` to specify the parameters. This allowed for specifying the behavior and values in the same object. For instance, setting the last parameter to true (in this patch) ignores several of the other parameters entirely which is weird. The code I was working with looks something like:

```c++
#include <boost/variant/variant_fwd.hpp>

namespace ssl
{
   struct none_
   {};
   constexpr const none_ none{};

   struct any_certificate_
   {};
   constexpr const any_certificate_ any_certificate{};

   struct default_certificates_
   {};
   constexpr const default_certificates_ default_certificates{};

    struct certificate_file
    {
        std::string path;
    };

   using mode = boost::variant<none_, any_certificate_, default_certificates_, certificate_file>;
}

// client
#include <boost/variant/variant.hpp>
m_http_client.connect(address, port, timeout, ssl::default_certificates);
m_http_client.connect(address, port, timeout, ssl::certificate_file{"/etc/server/file.pem"});
```

Also, if multiple certs are removed then it becomes much easier because there is built-in functionality with `asio::ssl`/`openssl` to use a single cert. Since one certificate can sign multiple server certificates, when would this functionality be used? It seems easier to omit the logic (and thus any potential bugs/memory leaks).

Reply:
OK, seems fair enough. The param list is a bit unwieldy indeed.

If I remove the multiple cert thing, then not much code is saved, and it forces the user to select the right cert for every run if they usually use more than one peer. Given it's C++ containers, I really do not expects leaks either.

Reply:
I meant the usage of openssl calls. `create_ssl_certificate` should need some frees at the end, and there are a number of places where a `std::string` constructor can throw and leak memory. The latter scenario is unlikely at least.

Reply:
Sure, but using this as an argument for supporting only one allowed certificate seems reaching quite a bit. If there are leaks on error paths, let's fix them.

Reply:
What code are you suggesting replacing ? The only code I see looping through allowed certificates is:

<pre>
  std::list<std::string> ssl_allowed_certificates;
  for (const std::string &path: daemon_ssl_allowed_certificates)
  {
      ssl_allowed_certificates.push_back({});
      if (!epee::file_io_utils::load_file_to_string(path, ssl_allowed_certificates.back()))
      {
        MERROR("Failed to load certificate: " << path);
        ssl_allowed_certificates.back() = std::string();
      }
  } 
</pre>

but that doesn't quite seem right to me.